src/node/mini_miner.cpp

```diff
@@ -0,0 +1,265 @@
+    // Anything otherwise unavailable just has a bump fee of 0
+    for (const auto& outpoint : outpoints) {
+        if (const auto ptx{mempool.GetConflictTx(outpoint)}) {
+            // The conflicting transaction is to-be-replaced
+            to_be_replaced.insert(ptx->GetHash());
+        }
+
+        if (!mempool.exists(GenTxid::Txid(outpoint.hash))) {
+            // This UTXO is either confirmed or not yet submitted to mempool.
+            // In the former case, no bump fee is required.
+            // In the latter case, we have no information, so just return 0.
+            this->bump_fees.emplace(std::make_pair(outpoint, 0));
+        } else {
+            // This UTXO is unconfirmed, in the mempool, and available to spend.
+            auto it = outpoints_needed_by_txid.find(outpoint.hash);
+            if (it != outpoints_needed_by_txid.end()) {
+                it->second.push_back(outpoint);
+            } else {
+                std::vector<COutPoint> outpoints_of_tx({outpoint});
+                outpoints_needed_by_txid.emplace(std::make_pair(outpoint.hash, outpoints_of_tx));
+                // Instead of operating on the entire mempool, just run the mining algorithm on the
+                // cluster of relevant transactions, which we'll store in mapModifiedTx.
+            }
+        }
+    }
+    // Calculate the cluster and construct the entry map.
+    std::vector<uint256> txids_needed;
+    std::transform(outpoints_needed_by_txid.cbegin(),
+                   outpoints_needed_by_txid.cend(),
+                   std::back_inserter(txids_needed),
+                   [](const auto& pair) { return pair.first; });
+    const auto& cluster = mempool.CalculateCluster(txids_needed);
+    for (const auto& txiter : cluster) {
+        if (to_be_replaced.find(txiter->GetTx().GetHash()) == to_be_replaced.end()) {
+            auto [mapiter, success] = entries_by_txid.emplace(std::make_pair(txiter->GetTx().GetHash(), MockMempoolEntry(txiter)));
+            assert(success);
+            entries.push_back(mapiter);
+        } else {
+            auto outpoints_it = outpoints_needed_by_txid.find(txiter->GetTx().GetHash());
+            if (outpoints_it != outpoints_needed_by_txid.end()) {
+                for (const auto& outpoint : outpoints_it->second) {
+                    this->bump_fees.emplace(std::make_pair(outpoint, 0));
+                }
+            }
+        }
+    }
+
+    // Remove the to-be-replaced transactions and build the descendant_set_by_txid cache.
+    for (const auto& txiter : cluster) {
+        const auto& txid = txiter->GetTx().GetHash();
+        // Cache descendants for future use. Unlike the real mempool, a descendant MockMempoolEntry
+        // will not exist without its ancestor MockMempoolEntry, so these sets won't be invalidated.
+        std::vector<MockEntryMap::iterator> cached_descendants;
+        cached_descendants.emplace_back(entries_by_txid.find(txid));
```

Comment:
Something more important: this `emplace_back()` shouldn't be here at all, because each mempool entry is a member of its own descendants' list. So this "main" transaction gets added to `cached_descendants` below. This was causing the `fee_with_ancestors` and `vsize_with_ancestors` to go negative near the end of `BuildMockTemplate()` below.